test/gtest/ucs/test_buffer.cc

```diff
@@ -0,0 +1,63 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2019.  ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
+
+#include <common/test.h>
+#include <common/test_helpers.h>
+
+#include <ucs/memory/buffer.h>
+
+class test_buffer : public ucs::test {
+protected:
+};
+
+UCS_TEST_F(test_buffer, copy_to_wc) {
+    uint64_t zero = 0;
+
+    for (size_t i = 1; i < 1024; i++) {
+        size_t tail   = ucs_padding(i, sizeof(uint64_t));
+        size_t d_size = ucs_align_up(i, sizeof(uint64_t));
+        char *host;
+        char *dm;
+
+        host = (char*)malloc(i);
+        dm   = (char*)malloc(d_size);
+
+        EXPECT_TRUE(host != NULL);
+        EXPECT_TRUE(dm != NULL);
+
+        ucs::fill_random(host, i);
+        ucs::fill_random(dm, d_size);
+
+        EXPECT_TRUE(ucs_buffer_copy_to_wc(dm, host, i) == dm);
+        EXPECT_TRUE(memcmp(host, dm, i) == 0);
```

Comment:
Who is responsible to order access to the memory ? Seems like SFENCE is missing ?

Reply:
There is more substantial question, if the sfence and a like should be part of memcpy routine. If not we would have to expose underpaying memory model to user, which gives me nightmares :) Also, I would like to get some feedback from GPU community, since it might be directly mapped to GPU and I’m not sure about nvlink memory semantics.

Also, I think we shell move it to experimental API.

Reply:
Under memcpy I mean the new routine to move from and to the device.

Reply:
user should call fence after all buffers are complete (or UCX will call fence on send operation). fence is expensive operation and should not be called on every memcopy

Reply:
@hoopoepg how user will know about this ? What if there is an architecture that does not require the fence ?